src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -542,7 +542,12 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
 
             if (!convertedExpression.HasErrors && !hasErrors)
             {
-                if (constantValueOpt == null)
+                var conversion = convertedExpression as BoundConversion;
+                if (conversion != null && conversion.ConversionKind == ConversionKind.ImplicitUserDefined)
+                {
+                    diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, conversion.Operand.Type!, conversion.Type);
```

Comment:
>conversion.Operand [](http://example.com/codeflow?start=118&length=18)

I think in some cases ```conversion.Operand``` this might not correspond to the syntactic operand for the conversion, Therefore, the type could be surprising and the error message would not make sense because, for example, there is a conversion that supports constant folding between the types.

Reply:
I have updated this to use `expression` instead - this should be the expression being bound and as I understand that does makes sense. Stepping through a few unit tests confirmed my beliefs. This change however, caused errors when the expression was a null literal so I have written a case to show ErrorCode.ERR_ValueCantBeNull when null literals are attempted to be converted to a non-nullable type.

For the following snipped: 
`static bool M1(Span<char> chars) => chars is null;`
This yields the following message:
`// (5,50): error CS0037: Cannot convert null to 'Span<char>' because it is a non-nullable value type` 
instead of 
`// (5,50): error CS0150: A constant value is expected`

Affected test cases were adjusted to this change.

Looking forward to hear your feedback on this change.

Reply:
What kind of conversion is actually used for `Span<char> x = null`?

Reply:
That appears to be an ImplicitReference conversion.

Reply:
> That appears to be an ImplicitReference conversion.

I started doubting the integrity of my recent change with respect to null literals and have taken another look. In the pattern expression the conversion is different, its actually implicit user defined. When there does not exist such a conversion the correct `CS0037: Cannot convert null to 'Span<char>' because it is a non-nullable value type` is already shown so I will proceed to remove the added check for that.

Now I wonder: when an implicit user defined conversion such as `static implicit operator Span<T>(T[]? array)` exists, then the error message in a pattern which attempts to convert null to Span would become `CS9098: Cannot implicitly convert type 'char[]' to type 'Span<char>' using user-defined conversion in pattern expression`. Would this be appropriate?